File: GenesisAndroid/src/test/java/com/emerchantpay/gateway/genesisandroid/validation/RequiredParametersValidatorUnitTest.java
package com.emerchantpay.gateway.genesisandroid.validation;

import android.content.Context;
import android.test.mock.MockContext;

import com.emerchantpay.gateway.genesisandroid.api.constants.KlarnaItemTypes;
import com.emerchantpay.gateway.genesisandroid.api.internal.request.PaymentRequest;
import com.emerchantpay.gateway.genesisandroid.api.internal.request.TransactionTypesRequest;
import com.emerchantpay.gateway.genesisandroid.api.internal.validation.RequiredParameters;
import com.emerchantpay.gateway.genesisandroid.api.internal.validation.RequiredParametersValidator;
import com.emerchantpay.gateway.genesisandroid.api.models.Country;
import com.emerchantpay.gateway.genesisandroid.api.models.Currency;
import com.emerchantpay.gateway.genesisandroid.api.models.PaymentAddress;
import com.emerchantpay.gateway.genesisandroid.api.models.WPFTransactionTypes;
import com.emerchantpay.gateway.genesisandroid.api.models.klarna.KlarnaItem;

import org.junit.Before;
import org.junit.Test;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.UUID;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class RequiredParametersValidatorUnitTest {

    // Application context
    private Context context;

    // Genesis Validator
    private RequiredParameters requiredParameters = new RequiredParameters();
    private RequiredParametersValidator validator;

    // Payment request
    private PaymentRequest request;

    // Payment address
    private PaymentAddress billingAddress;


    // Transaction types
    private TransactionTypesRequest transactionTypes;

    // Parameters
    private String transactionId;
    private BigDecimal amount;
    private String customerEmail;
    private String customerPhone;
    private String notificationUrl;

    private HashMap<String, String> requiredParamsMap;

    @Before
    public void setupValidData() throws IllegalAccessException {
        context = new MockContext().getApplicationContext();

        // Intitial params
        transactionId = UUID.randomUUID().toString();
        amount = new BigDecimal("2.00");
        customerEmail = "johndoe@example.com";
        customerPhone = "+555555555";
        notificationUrl = "http://google.com";

        // Address
        billingAddress = new PaymentAddress("John", "Doe",
                "address1", "", "10000", "New York",
                "state", new Country().getCountry("United States"));

        // Transaction types list
        transactionTypes = new TransactionTypesRequest();
        transactionTypes.addTransaction(WPFTransactionTypes.authorize);
        transactionTypes.addTransaction(WPFTransactionTypes.ezeewallet);

        // Payment request
        request = new PaymentRequest(context, transactionId, amount, Currency.USD,
                customerEmail, customerPhone, billingAddress, notificationUrl, transactionTypes);
    }


    // Request params
    @Test
    public void testRequestParams() {
        HashMap<String, String> map = requiredParameters.getRequiredParametersForRequest(request);

        validator = new RequiredParametersValidator(map);
        assertTrue(validator.isValidRequiredParams());
    }

    @Test
    public void testWithMissingRequestParams() throws IllegalAccessException {
        // Intitial params
        transactionId = UUID.randomUUID().toString();
        amount = new BigDecimal("2.00");
        customerEmail = null;
        customerPhone = "+555555555";
        notificationUrl = "http://google.com";

        // Payment request
        request = new PaymentRequest(context, transactionId, amount, Currency.USD,
                customerEmail, customerPhone, billingAddress, notificationUrl, transactionTypes);

        requiredParamsMap = requiredParameters.getRequiredParametersForRequest(request);

        validator = new RequiredParametersValidator(requiredParamsMap);
        assertFalse(validator.isValidRequiredParams());
    }

    // Billing address params
    @Test
    public void testAddressParams() {
        HashMap<String, String> map = requiredParameters.getRequiredParametersForAddress(billingAddress);

        validator = new RequiredParametersValidator(map);
        assertTrue(validator.isValidRequiredParams());
    }

    @Test
    public void testWithMissingAddressParams() throws IllegalAccessException {
        // Address
        billingAddress = new PaymentAddress("John", "Doe",
                null, "", "10000", "New York",
                "state", new Country().getCountry("United States"));

        requiredParamsMap = requiredParameters.getRequiredParametersForAddress(billingAddress);

        validator = new RequiredParametersValidator(requiredParamsMap);
        assertFalse(validator.isValidRequiredParams());
    }

    // Transaction types params
    @Test
    public void testWithTransactionTypesParams() throws IllegalAccessException {
        // Transaction types list
        transactionTypes = new TransactionTypesRequest();
        transactionTypes.addTransaction(WPFTransactionTypes.ppro);
        transactionTypes.addParam("product_name", "TICKETS");

        // Payment request
        request = new PaymentRequest(context, transactionId, amount, Currency.USD,
                customerEmail, customerPhone, billingAddress, notificationUrl, transactionTypes);

        for (String transactionType : transactionTypes.getTransactionTypesList()) {
            HashMap<String, String> map = requiredParameters.getRequiredParametersForTransactionType(request, transactionType);

            validator = new RequiredParametersValidator(map);

            assertTrue(validator.isValidRequiredParams());
        }
    }

    @Test
    public void testWithMissingTransactionTypesParams() throws IllegalAccessException {
        // Transaction types list
        transactionTypes = new TransactionTypesRequest();
        transactionTypes.addTransaction(WPFTransactionTypes.ppro);

        // Payment request
        request = new PaymentRequest(context, transactionId, amount, Currency.USD,
                customerEmail, customerPhone, billingAddress, notificationUrl, transactionTypes);

        HashMap<String, String> map;

        for (String transactionType : transactionTypes.getTransactionTypesList()) {
            map = requiredParameters.getRequiredParametersForTransactionType(request, transactionType);

            validator = new RequiredParametersValidator(map);

            assertFalse(validator.isValidRequiredParams());
        }
    }

    // Klarna params
    @Test
    public void testWithKlarnaItems() {
        KlarnaItem item = new KlarnaItem("TICKET", KlarnaItemTypes.DISCOUNT, 10,
                new BigDecimal(10.00), new BigDecimal(2.00));

        request.addKlarnaItem(item);

        requiredParamsMap = requiredParameters.getRequiredParametersForKlarnaItem(request.getKlarnaItemsRequest());

        validator = new RequiredParametersValidator(requiredParamsMap);
        assertTrue(validator.isValidRequiredParams());
    }

    @Test
    public void testWithMissingKlarnaItems() {
        KlarnaItem item = new KlarnaItem(null, KlarnaItemTypes.DISCOUNT, 10,
                new BigDecimal(10.00), new BigDecimal(2.00));

        request.addKlarnaItem(item);

        HashMap<String, String> map = requiredParameters.getRequiredParametersForKlarnaItem(request.getKlarnaItemsRequest());

        validator = new RequiredParametersValidator(map);
        assertFalse(validator.isValidRequiredParams());
    }
}

File: GenesisAndroid/src/test/java/com/emerchantpay/gateway/genesisandroid/validation/RequiredParametersValidatorUnitTest.kt
package com.emerchantpay.gateway.genesisandroid.validation

import android.content.Context
import com.emerchantpay.gateway.genesisandroid.api.constants.KlarnaItemTypes
import com.emerchantpay.gateway.genesisandroid.api.constants.WPFTransactionTypes
import com.emerchantpay.gateway.genesisandroid.api.internal.request.PaymentRequest
import com.emerchantpay.gateway.genesisandroid.api.internal.request.TransactionTypesRequest
import com.emerchantpay.gateway.genesisandroid.api.internal.validation.RequiredParameters
import com.emerchantpay.gateway.genesisandroid.api.internal.validation.RequiredParametersValidator
import com.emerchantpay.gateway.genesisandroid.api.models.Country
import com.emerchantpay.gateway.genesisandroid.api.models.Currency
import com.emerchantpay.gateway.genesisandroid.api.models.PaymentAddress
import com.emerchantpay.gateway.genesisandroid.api.models.klarna.KlarnaItem
import io.mockk.mockk
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.math.BigDecimal
import java.util.*

class RequiredParametersValidatorUnitTest {

    // Application context
    private var context: Context? = null

    // Genesis Validator
    private val requiredParameters = RequiredParameters()
    private var validator: RequiredParametersValidator? = null

    // Payment request
    private var request: PaymentRequest? = null

    // Payment address
    private var billingAddress: PaymentAddress? = null


    // Transaction types
    private var transactionTypes: TransactionTypesRequest? = null

    // Parameters
    private var transactionId: String? = null
    private var amount: BigDecimal? = null
    private var customerEmail: String? = null
    private var customerPhone: String? = null
    private var notificationUrl: String? = null

    private var requiredParamsMap: HashMap<String, String>? = null

    @Before
    @Throws(IllegalAccessException::class)
    fun setupValidData() {
        // Intitial params
        context = mockk<Context>(relaxed = true)
        transactionId = UUID.randomUUID().toString()
        amount = BigDecimal("2.00")
        customerEmail = "johndoe@example.com"
        customerPhone = "+555555555"
        notificationUrl = "http://google.com"

        // Address
        billingAddress = PaymentAddress("John", "Doe",
                "address1", "", "10000", "New York",
                "state", Country().getCountry("United States")!!)

        // Transaction types list
        transactionTypes = TransactionTypesRequest()
        transactionTypes!!.addTransaction(WPFTransactionTypes.AUTHORIZE)
        transactionTypes!!.addTransaction(WPFTransactionTypes.EZEEWALLET)

        // Payment request
        request = context?.let {
            PaymentRequest(it, transactionId!!, amount!!, Currency.USD,
                    customerEmail!!, customerPhone!!, billingAddress!!, notificationUrl, transactionTypes!!)
        }

        request!!.consumerId = "123456"
    }


    // Request params
    @Test
    fun testRequestParams() {
        val map = request?.let { requiredParameters.getRequiredParametersForRequest(it) }

        validator = RequiredParametersValidator(map)
        assertTrue(validator!!.isValidRequiredParams!!)
    }

    @Test
    @Throws(IllegalAccessException::class)
    fun testWithMissingRequestParams() {
        // Intitial params
        transactionId = UUID.randomUUID().toString()
        amount = BigDecimal("2.00")
        customerEmail = ""
        customerPhone = ""
        notificationUrl = "http://google.com"

        // Payment request
        request = context?.let {
            billingAddress?.let { it1 ->
                transactionTypes?.let { it2 ->
                    PaymentRequest(it, transactionId, amount, Currency.USD,
                            customerEmail, customerPhone, it1, notificationUrl, it2)
                }
            }
        }

        requiredParamsMap = request?.let { requiredParameters.getRequiredParametersForRequest(it) }

        validator = RequiredParametersValidator(requiredParamsMap)
        assertTrue(validator!!.isValidRequiredParams!!)
    }

    @Test
    @Throws(IllegalAccessException::class)
    fun testWithNullRequestParams() {
        // Intitial params
        transactionId = UUID.randomUUID().toString()
        amount = BigDecimal("2.00")
        notificationUrl = "http://google.com"

        // Payment request
        request = context?.let {
            billingAddress?.let { it1 ->
                transactionTypes?.let { it2 ->
                    PaymentRequest(it, transactionId, amount, Currency.USD,
                        null, null, it1, notificationUrl, it2)
                }
            }
        }

        requiredParamsMap = request?.let { requiredParameters.getRequiredParametersForRequest(it) }

        validator = RequiredParametersValidator(requiredParamsMap)
        assertTrue(validator!!.isValidRequiredParams!!)
    }

    // Billing address params
    @Test
    fun testAddressParams() {
        val map = billingAddress?.let { requiredParameters.getRequiredParametersForAddress(it) }

        validator = RequiredParametersValidator(map)
        assertTrue(validator!!.isValidRequiredParams!!)
    }

    @Test
    @Throws(IllegalAccessException::class)
    fun testWithMissingAddressParams() {
        // Address
        billingAddress = PaymentAddress("John", "Doe",
                "", "", "", "",
                "", Country().getCountry("United States")!!)

        requiredParamsMap = requiredParameters.getRequiredParametersForAddress(billingAddress!!)

        validator = RequiredParametersValidator(requiredParamsMap)
        assertFalse(validator!!.isValidRequiredParams!!)
    }

    // Transaction types params
    @Test
    @Throws(IllegalAccessException::class)
    fun testWithTransactionTypesParams() {
        // Transaction types list
        transactionTypes = TransactionTypesRequest()
        transactionTypes!!.addTransaction(WPFTransactionTypes.PPRO)
        transactionTypes!!.addParam("product_name", "TICKETS")

        // Payment request
        request = billingAddress?.let {
            context?.let { it1 ->
                transactionId?.let { it2 ->
                    amount?.let { it3 ->
                        customerEmail?.let { it4 ->
                            customerPhone?.let { it5 ->
                                PaymentRequest(it1, it2, it3, Currency.USD,
                                        it4, it5, it, notificationUrl, transactionTypes!!)
                            }
                        }
                    }
                }
            }
        }

        for (transactionType in transactionTypes!!.transactionTypesList) {
            val map = request?.let { requiredParameters.getRequiredParametersForTransactionType(it, transactionType) }

            validator = RequiredParametersValidator(map)

            assertTrue(validator!!.isValidRequiredParams!!)
        }
    }

    @Test
    @Throws(IllegalAccessException::class)
    fun testWithMissingTransactionTypesParams() {
        // Transaction types list
        transactionTypes = TransactionTypesRequest()
        transactionTypes!!.addTransaction(WPFTransactionTypes.PPRO)

        // Payment request
        request = PaymentRequest(context, transactionId, amount, Currency.USD,
                customerEmail, customerPhone, billingAddress, notificationUrl, transactionTypes!!)

        var map: HashMap<String, String>

        transactionTypes!!.transactionTypesList.forEach { transactionType ->
            map = requiredParameters.getRequiredParametersForTransactionType(request!!, transactionType)

            validator = RequiredParametersValidator(map)

            assertFalse(validator!!.isValidRequiredParams!!)
        }
    }

    @Test
    @Throws(IllegalAccessException::class)
    fun testTokenization() {
        // Transaction types list
        transactionTypes = TransactionTypesRequest()
        transactionTypes!!.addTransaction(WPFTransactionTypes.PPRO)

        // Payment request
        request = PaymentRequest(context, transactionId, amount, Currency.USD,
                customerEmail, customerPhone, billingAddress, notificationUrl, transactionTypes!!)

        request!!.rememberCard = true

        var map: HashMap<String, String>

        transactionTypes!!.transactionTypesList.forEach { transactionType ->
            map = requiredParameters.getRequiredParametersForTransactionType(request!!, transactionType)

            validator = RequiredParametersValidator(map)

            assertFalse(validator!!.isValidRequiredParams!!)
        }
    }

    // Klarna params
    @Test
    fun testWithKlarnaItems() {
        val item = KlarnaItem("TICKET", KlarnaItemTypes.DISCOUNT, 10,
                BigDecimal(10.00), BigDecimal(2.00))

        request!!.addKlarnaItem(item)

        requiredParamsMap = request!!.klarnaItemsRequest?.let { requiredParameters.getRequiredParametersForKlarnaItem(it) }

        validator = RequiredParametersValidator(requiredParamsMap)
        assertTrue(validator!!.isValidRequiredParams!!)
    }

    @Test
    fun testWithMissingKlarnaItems() {
        val item = KlarnaItem("", KlarnaItemTypes.DISCOUNT, 10,
                BigDecimal(10.00), BigDecimal(2.00))

        request!!.addKlarnaItem(item)

        val map = request!!.klarnaItemsRequest?.let { requiredParameters.getRequiredParametersForKlarnaItem(it) }

        validator = RequiredParametersValidator(map)
        assertFalse(validator!!.isValidRequiredParams!!)
    }
}